From 0460b58b5efd4450e074a39227dc9a9fd589ccad Mon Sep 17 00:00:00 2001
From: Oliver Beard <olib141@outlook.com>
Date: Mon, 20 Oct 2025 12:57:22 +0100
Subject: [PATCH] frontend/wallpaper: Adapt to WallpaperItem config changes

This commit fixes the wallpaper by adapting to changes in how the wallpaper loads configuration in libplasma@c7e0ba9b.

BUG: 507842
---
 src/frontend/wallpaper/wallpaperapp.cpp | 49 +++++++++++++++++--------
 src/frontend/wallpaper/wallpaperapp.h   |  2 -
 2 files changed, 33 insertions(+), 18 deletions(-)

diff --git a/src/frontend/wallpaper/wallpaperapp.cpp b/src/frontend/wallpaper/wallpaperapp.cpp
index 0f91836a..e09fe700 100644
--- a/src/frontend/wallpaper/wallpaperapp.cpp
+++ b/src/frontend/wallpaper/wallpaperapp.cpp
@@ -6,38 +6,28 @@
     SPDX-License-Identifier: GPL-2.0-or-later
 */
 
+#include <QFile>
+#include <QQmlContext>
+#include <QQuickItem>
+
+#include <KConfigLoader>
+#include <KConfigPropertyMap>
 #include <KPackage/PackageLoader>
 #include <KWindowSystem>
 
 #include <LayerShellQt/Shell>
 
 #include "plasmaloginsettings.h"
-#include "wallpaperintegration.h"
 
 #include "wallpaperwindow.h"
 
 #include "wallpaperapp.h"
 
-class WallpaperItem : public WallpaperIntegration
-{
-    Q_OBJECT
-public:
-    explicit WallpaperItem(QQuickItem *parent = nullptr)
-        : WallpaperIntegration(parent)
-    {
-        setConfig(PlasmaLoginSettings::getInstance().sharedConfig());
-        setPluginName(PlasmaLoginSettings::getInstance().wallpaperPluginId());
-        init();
-    }
-};
-
 WallpaperApp::WallpaperApp(int &argc, char **argv)
     : QGuiApplication(argc, argv)
 {
     LayerShellQt::Shell::useLayerShell();
 
-    qmlRegisterType<WallpaperItem>("org.kde.plasma.plasmoid", 2, 0, "WallpaperItem");
-
     m_wallpaperPackage = KPackage::PackageLoader::self()->loadPackage(QStringLiteral("Plasma/Wallpaper"));
     m_wallpaperPackage.setPath(PlasmaLoginSettings::getInstance().wallpaperPluginId());
 
@@ -64,6 +54,10 @@ void WallpaperApp::adoptScreen(QScreen *screen)
     window->setVisible(true);
     m_windows << window;
 
+    connect(screen, &QScreen::geometryChanged, this, [window](const QRect &geometry) {
+        window->setGeometry(geometry);
+    });
+
     connect(screen, &QObject::destroyed, window, [this, window]() {
         m_windows.removeAll(window);
         window->deleteLater();
@@ -79,7 +73,30 @@ void WallpaperApp::setupWallpaperPlugin(WallpaperWindow *window)
         return;
     }
 
+    const QString xmlPath = m_wallpaperPackage.filePath(QByteArrayLiteral("config"), QStringLiteral("main.xml"));
+
+    const KConfigGroup cfg = PlasmaLoginSettings::getInstance()
+                                 .sharedConfig()
+                                 ->group(QStringLiteral("Greeter"))
+                                 .group(QStringLiteral("Wallpaper"))
+                                 .group(PlasmaLoginSettings::getInstance().wallpaperPluginId());
+
+    KConfigLoader *configLoader;
+    if (xmlPath.isEmpty()) {
+        configLoader = new KConfigLoader(cfg, nullptr, this);
+    } else {
+        QFile file(xmlPath);
+        configLoader = new KConfigLoader(cfg, &file, this);
+    }
+
+    KConfigPropertyMap *config = new KConfigPropertyMap(configLoader, this);
+    // potd (picture of the day) is using a kded to monitor changes and
+    // cache data for the lockscreen. Let's notify it.
+    config->setNotify(true);
+
     window->setSource(QUrl::fromLocalFile(m_wallpaperPackage.filePath("mainscript")));
+    window->rootObject()->setProperty("configuration", QVariant::fromValue(config));
+    window->rootObject()->setProperty("pluginName", PlasmaLoginSettings::getInstance().wallpaperPluginId());
 }
 
 #include "wallpaperapp.moc"
diff --git a/src/frontend/wallpaper/wallpaperapp.h b/src/frontend/wallpaper/wallpaperapp.h
index d771d99a..8d63713f 100644
--- a/src/frontend/wallpaper/wallpaperapp.h
+++ b/src/frontend/wallpaper/wallpaperapp.h
@@ -18,8 +18,6 @@ namespace PlasmaQuick
 class QuickViewSharedEngine;
 }
 
-class WallpaperIntegration;
-
 class WallpaperWindow;
 
 class WallpaperApp : public QGuiApplication
-- 
GitLab